wfe2/wfe.go

```diff
@@ -695,22 +702,65 @@ func (wfe *WebFrontEndImpl) processRevocation(
 	// Compute and record the serial number of the provided certificate
 	serial := core.SerialToString(providedCert.SerialNumber)
 	logEvent.Extra["ProvidedCertificateSerial"] = serial
+	notFoundProb := probs.NotFound("No such certificate")
+
+	var certDER []byte
+	// If the PrecertificateRevocation feature flag is enabled that means we won't
+	// do a byte-for-byte comparison of the providedCert against the stored cert
+	// returned by SA.GetCertificate because a precert will always fail this
+	// check. Instead, perform a signature validation of the providedCert using
+	// the known issuer public keys. If the providedCert signature can not be
+	// validated with any of the known issuers return a not-found error.
+	if features.Enabled(features.PrecertificateRevocation) {
+		// If no issuerCertificates are initialized but the PrecertificateRevocation
+		// feature flag is enabled then return a runtime server internal error
+		// rather than fail open.
+		if len(wfe.issuerCertificates) == 0 {
+			return probs.ServerInternal(
+				"unable to verify provided certificate, empty issuerCertificates")
+		}
 
-	// Lookup the certificate by the serial. If the certificate wasn't found, or
-	// it wasn't a byte-for-byte match to the certificate requested for
-	// revocation, return an error
-	cert, err := wfe.SA.GetCertificate(ctx, serial)
-	if err != nil || !bytes.Equal(cert.DER, revokeRequest.CertificateDER) {
-		return probs.NotFound("No such certificate")
+		// Try to validate the signature on the provided cert using each of the
+		// known issuer certificates. This is O(n) but we always expect to have
```

Comment:
As a note, we could potentially rearrange the `certificateChains` structure so each entry would have the `[]byte` of the premade PEM chain, *plus* an `*x509.Certificate` for the issuer; then we could look this up in O(1) time. I don't think it's worth the time to rewrite this now, since as you say we expect the issuer count to be low.